File: fetch_region_boundaries.mjs
#!/usr/bin/env node
/**
 * fetch_region_boundaries.mjs
 * Geocodes + fetches polygon boundaries for all regions via Nominatim.
 * Uses upsert_region_boundary RPC to insert/update geographic_boundaries.
 *
 * Usage:
 *   node fetch_region_boundaries.mjs              # live run
 *   node fetch_region_boundaries.mjs --dry-run    # preview only
 *   node fetch_region_boundaries.mjs --limit 10   # process first N
 */

import { createClient } from '@supabase/supabase-js';
import { readFileSync } from 'fs';

const env = Object.fromEntries(
  readFileSync('frontend/.env', 'utf8')
    .split('\n')
    .filter(l => l.includes('='))
    .map(l => { const [k, ...v] = l.split('='); return [k.trim(), v.join('=').trim()] })
);

const supabase = createClient(env.VITE_SUPABASE_URL, env.VITE_SUPABASE_ANON_KEY);

const DRY_RUN = process.argv.includes('--dry-run');
const limitIdx = process.argv.indexOf('--limit');
const LIMIT = limitIdx !== -1 ? parseInt(process.argv[limitIdx + 1]) : Infinity;

// Haversine distance in km
function haversine(lat1, lng1, lat2, lng2) {
  const R = 6371;
  const dLat = (lat2 - lat1) * Math.PI / 180;
  const dLng = (lng2 - lng1) * Math.PI / 180;
  const a = Math.sin(dLat/2)**2 + Math.cos(lat1*Math.PI/180) * Math.cos(lat2*Math.PI/180) * Math.sin(dLng/2)**2;
  return R * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1-a));
}

// Compute centroid from GeoJSON polygon
function computeCentroid(geojson) {
  let totalLat = 0, totalLng = 0, count = 0;
  function extractCoords(coords) {
    if (typeof coords[0] === 'number') {
      totalLng += coords[0];
      totalLat += coords[1];
      count++;
    } else {
      coords.forEach(c => extractCoords(c));
    }
  }
  if (geojson.type === 'GeometryCollection') {
    geojson.geometries.forEach(g => extractCoords(g.coordinates));
  } else {
    extractCoords(geojson.coordinates);
  }
  return count > 0 ? { lat: totalLat / count, lng: totalLng / count } : null;
}

async function nominatimSearch(query, retries = 3) {
  for (let attempt = 0; attempt < retries; attempt++) {
    const url = `https://nominatim.openstreetmap.org/search?q=${encodeURIComponent(query)}&format=json&polygon_geojson=1&limit=1`;
    const res = await fetch(url, {
      headers: { 'User-Agent': 'LoamWineApp/1.0 (contact@loam.wine)' }
    });
    if (res.status === 429) {
      const wait = 5000 * (attempt + 1);
      process.stdout.write(`[rate-limited, waiting ${wait/1000}s] `);
      await sleep(wait);
      continue;
    }
    if (!res.ok) throw new Error(`HTTP ${res.status}`);
    return res.json();
  }
  throw new Error('Rate limited after retries');
}

async function sleep(ms) { return new Promise(r => setTimeout(r, ms)); }

async function main() {
  console.log(`\n=== Region Boundary Fetcher ===`);
  console.log(`Mode: ${DRY_RUN ? 'DRY RUN' : 'LIVE'}`);

  // Fetch all regions with their country name and existing boundary status
  const { data: regions, error } = await supabase
    .from('regions')
    .select('id, name, country:countries(id, name)')
    .order('name');

  if (error) { console.error('Failed to fetch regions:', error); process.exit(1); }

  // Get existing boundary rows for regions
  const { data: existing } = await supabase
    .from('geographic_boundaries')
    .select('id, region_id')
    .not('region_id', 'is', null);

  const existingMap = new Map((existing || []).map(e => [e.region_id, e.id]));

  // Get average appellation centroids per region for validation
  const centroidMap = new Map();
  const { data: bpData } = await supabase.rpc('get_boundary_points');
  if (bpData) {
    // Group appellation points by region_name + country_name
    const regionPoints = {};
    for (const pt of bpData) {
      if (pt.entity_type !== 'appellation' || !pt.region_name || !pt.lat || !pt.lng) continue;
      const key = `${pt.region_name}|${pt.country_name}`;
      if (!regionPoints[key]) regionPoints[key] = { sumLat: 0, sumLng: 0, count: 0 };
      regionPoints[key].sumLat += pt.lat;
      regionPoints[key].sumLng += pt.lng;
      regionPoints[key].count++;
    }
    // Map back to region IDs
    for (const r of regions) {
      const key = `${r.name}|${r.country?.name}`;
      if (regionPoints[key]) {
        const p = regionPoints[key];
        centroidMap.set(r.id, { lat: p.sumLat / p.count, lng: p.sumLng / p.count });
      }
    }
    console.log(`Computed expected centroids for ${centroidMap.size} regions from appellation data`);
  }

  // Filter to regions without polygon boundaries (keep those that are centroid-only or missing entirely)
  const todo = regions.filter(r => {
    // Skip regions where name == country name (catch-all regions)
    if (r.name === r.country?.name) return false;
    return true;
  });

  // Further filter: only those without existing boundaries, OR existing but we want to try polygons
  const toProcess = todo.filter(r => !existingMap.has(r.id));

  console.log(`Total regions: ${regions.length}`);
  console.log(`Skipping ${regions.length - todo.length} catch-all regions (name == country)`);
  console.log(`Already have boundaries: ${existingMap.size}`);
  console.log(`To process: ${toProcess.length}`);

  const limit = Math.min(toProcess.length, LIMIT);
  let success = 0, failed = 0, stored = 0;
  const countryCounts = {};

  for (let i = 0; i < limit; i++) {
    const r = toProcess[i];
    const countryName = r.country?.name || 'Unknown';
    process.stdout.write(`[${i+1}/${limit}] ${countryName}/${r.name}... `);

    await sleep(3000); // Rate limit (conservative to avoid 429s after heavy usage)

    // Build search queries - try multiple strategies
    const queries = [
      `${r.name}, ${countryName}`,
      r.name
    ];

    let found = false;
    for (const query of queries) {
      try {
        const results = await nominatimSearch(query);
        if (!results.length) continue;

        const result = results[0];
        const lat = parseFloat(result.lat);
        const lng = parseFloat(result.lon);
        const geojson = result.geojson;
        const hasPolygon = geojson && ['Polygon', 'MultiPolygon'].includes(geojson.type);

        // Validate: check if result is near the expected location (if we have appellation centroids)
        const expected = centroidMap.get(r.id);
        if (expected) {
          const dist = haversine(lat, lng, expected.lat, expected.lng);
          if (dist > 200) { // Regions can be larger, so 200km threshold
            process.stdout.write(`✗ too far (${Math.round(dist)}km, query: "${query}")\n `);
            continue;
          }
        }

        if (hasPolygon) {
          const centroid = computeCentroid(geojson);
          if (centroid && expected) {
            const polyDist = haversine(centroid.lat, centroid.lng, expected.lat, expected.lng);
            if (polyDist > 300) {
              process.stdout.write(`✗ polygon too far (${Math.round(polyDist)}km)\n `);
              continue;
            }
          }
        }

        if (!DRY_RUN) {
          if (hasPolygon) {
            const { error: rpcError } = await supabase.rpc('upsert_region_boundary', {
              p_region_id: r.id,
              p_geojson: JSON.stringify(geojson),
              p_source_id: `nominatim/${result.osm_type}/${result.osm_id}`,
              p_confidence: 'approximate'
            });
            if (rpcError) {
              console.log(`✗ DB error: ${rpcError.message}`);
              failed++;
              found = true;
              break;
            }
          } else {
            // Centroid only
            const { error: rpcError } = await supabase.rpc('upsert_region_boundary', {
              p_region_id: r.id,
              p_lat: lat,
              p_lng: lng,
              p_source_id: `nominatim/${result.osm_type}/${result.osm_id}`,
              p_confidence: 'geocoded'
            });
            if (rpcError) {
              console.log(`✗ DB error: ${rpcError.message}`);
              failed++;
              found = true;
              break;
            }
          }
        }

        const typeLabel = hasPolygon ? `${geojson.type}` : 'centroid';
        console.log(`✓ ${typeLabel} (query: "${query}")`);
        success++;
        stored++;
        countryCounts[countryName] = (countryCounts[countryName] || 0) + 1;
        found = true;
        break;

      } catch (e) {
        process.stdout.write(`✗ error: ${e.message}\n `);
        await sleep(2000);
      }
    }

    if (!found) {
      console.log(`✗ no result found`);
      failed++;
    }
  }

  console.log(`\n=== Summary ===`);
  console.log(`Success: ${success}  Failed: ${failed}`);
  console.log(`\nStored (${stored}):`);
  Object.entries(countryCounts)
    .sort((a, b) => b[1] - a[1])
    .forEach(([c, n]) => console.log(`  ${c}: ${n}`));
}

main().catch(console.error);

File: pipeline/geo/fetch_region_boundaries.py
"""
Geocodes + fetches polygon boundaries for all regions via Nominatim.
Uses upsert_region_boundary RPC to insert/update geographic_boundaries.

Usage:
  python -m pipeline.geo.fetch_region_boundaries              # live run
  python -m pipeline.geo.fetch_region_boundaries --dry-run    # preview only
  python -m pipeline.geo.fetch_region_boundaries --limit 10
"""

import sys
import json
import math
import time
import argparse
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[2]))

import httpx
from pipeline.lib.db import get_supabase

USER_AGENT = "LoamWineApp/1.0 (contact@loam.wine)"


def haversine(lat1: float, lng1: float, lat2: float, lng2: float) -> float:
    R = 6371
    d_lat = (lat2 - lat1) * math.pi / 180
    d_lng = (lng2 - lng1) * math.pi / 180
    a = (math.sin(d_lat / 2) ** 2 +
         math.cos(lat1 * math.pi / 180) * math.cos(lat2 * math.pi / 180) *
         math.sin(d_lng / 2) ** 2)
    return R * 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))


def compute_centroid(geojson: dict) -> dict | None:
    total_lat = 0
    total_lng = 0
    count = 0

    def extract_coords(coords):
        nonlocal total_lat, total_lng, count
        if isinstance(coords[0], (int, float)):
            total_lng += coords[0]
            total_lat += coords[1]
            count += 1
        else:
            for c in coords:
                extract_coords(c)

    if geojson.get("type") == "GeometryCollection":
        for g in geojson.get("geometries", []):
            extract_coords(g["coordinates"])
    else:
        extract_coords(geojson["coordinates"])

    return {"lat": total_lat / count, "lng": total_lng / count} if count > 0 else None


def nominatim_search(client: httpx.Client, query: str, retries: int = 3) -> list[dict]:
    for attempt in range(retries):
        resp = client.get(
            "https://nominatim.openstreetmap.org/search",
            params={"q": query, "format": "json", "polygon_geojson": "1", "limit": "1"},
            headers={"User-Agent": USER_AGENT},
        )
        if resp.status_code == 429:
            wait = 5 * (attempt + 1)
            print(f"[rate-limited, waiting {wait}s] ", end="", flush=True)
            time.sleep(wait)
            continue
        resp.raise_for_status()
        return resp.json()
    raise Exception("Rate limited after retries")


def main():
    parser = argparse.ArgumentParser(description="Fetch region boundary polygons via Nominatim")
    parser.add_argument("--dry-run", action="store_true")
    parser.add_argument("--limit", type=int, default=None)
    args = parser.parse_args()

    sb = get_supabase()
    print(f"\n=== Region Boundary Fetcher ===")
    print(f"Mode: {'DRY RUN' if args.dry_run else 'LIVE'}")

    # Fetch all regions with country name
    all_regions = []
    offset = 0
    while True:
        result = sb.table("regions").select(
            "id, name, country:countries(id, name)"
        ).order("name").range(offset, offset + 999).execute()
        all_regions.extend(result.data)
        if len(result.data) < 1000:
            break
        offset += 1000

    # Get existing boundary rows for regions
    existing = []
    offset = 0
    while True:
        result = sb.table("geographic_boundaries").select("id, region_id").not_(
            "region_id", "is", "null"
        ).range(offset, offset + 999).execute()
        existing.extend(result.data)
        if len(result.data) < 1000:
            break
        offset += 1000
    existing_map = {e["region_id"]: e["id"] for e in existing}

    # Get average appellation centroids per region for validation
    centroid_map = {}
    bp_result = sb.rpc("get_boundary_points").execute()
    bp_data = bp_result.data or []
    if bp_data:
        region_points = {}
        for pt in bp_data:
            if pt.get("entity_type") != "appellation" or not pt.get("region_name"):
                continue
            if not pt.get("lat") or not pt.get("lng"):
                continue
            key = f"{pt['region_name']}|{pt['country_name']}"
            if key not in region_points:
                region_points[key] = {"sum_lat": 0, "sum_lng": 0, "count": 0}
            region_points[key]["sum_lat"] += pt["lat"]
            region_points[key]["sum_lng"] += pt["lng"]
            region_points[key]["count"] += 1

        for r in all_regions:
            country_name = r.get("country", {}).get("name", "")
            key = f"{r['name']}|{country_name}"
            if key in region_points:
                p = region_points[key]
                centroid_map[r["id"]] = {
                    "lat": p["sum_lat"] / p["count"],
                    "lng": p["sum_lng"] / p["count"],
                }
        print(f"Computed expected centroids for {len(centroid_map)} regions from appellation data")

    # Filter to non-catch-all regions without existing boundaries
    todo = [r for r in all_regions if r["name"] != r.get("country", {}).get("name")]
    to_process = [r for r in todo if r["id"] not in existing_map]

    print(f"Total regions: {len(all_regions)}")
    print(f"Skipping {len(all_regions) - len(todo)} catch-all regions")
    print(f"Already have boundaries: {len(existing_map)}")
    print(f"To process: {len(to_process)}")

    limit = min(len(to_process), args.limit) if args.limit else len(to_process)
    success = 0
    failed = 0
    stored = 0
    country_counts = {}

    client = httpx.Client(timeout=30)

    for i in range(limit):
        r = to_process[i]
        country_name = r.get("country", {}).get("name", "Unknown")
        print(f"[{i + 1}/{limit}] {country_name}/{r['name']}... ", end="", flush=True)

        time.sleep(3)  # Conservative rate limit

        queries = [f"{r['name']}, {country_name}", r["name"]]
        found = False

        for query in queries:
            try:
                results = nominatim_search(client, query)
                if not results:
                    continue

                result = results[0]
                lat = float(result["lat"])
                lng = float(result["lon"])
                geojson = result.get("geojson")
                has_poly = geojson and geojson.get("type") in ("Polygon", "MultiPolygon")

                # Validate distance
                expected = centroid_map.get(r["id"])
                if expected:
                    dist = haversine(lat, lng, expected["lat"], expected["lng"])
                    if dist > 200:
                        print(f"REJECT too far ({round(dist)}km, query: \"{query}\")")
                        continue

                if has_poly:
                    centroid = compute_centroid(geojson)
                    if centroid and expected:
                        poly_dist = haversine(centroid["lat"], centroid["lng"],
                                              expected["lat"], expected["lng"])
                        if poly_dist > 300:
                            print(f"REJECT polygon too far ({round(poly_dist)}km)")
                            continue

                if not args.dry_run:
                    source_id = f"nominatim/{result.get('osm_type')}/{result.get('osm_id')}"
                    if has_poly:
                        sb.rpc("upsert_region_boundary", {
                            "p_region_id": r["id"],
                            "p_geojson": json.dumps(geojson),
                            "p_source_id": source_id,
                            "p_confidence": "approximate",
                        }).execute()
                    else:
                        sb.rpc("upsert_region_boundary", {
                            "p_region_id": r["id"],
                            "p_lat": lat, "p_lng": lng,
                            "p_source_id": source_id,
                            "p_confidence": "geocoded",
                        }).execute()

                type_label = geojson["type"] if has_poly else "centroid"
                print(f"ok {type_label} (query: \"{query}\")")
                success += 1
                stored += 1
                country_counts[country_name] = country_counts.get(country_name, 0) + 1
                found = True
                break
            except Exception as e:
                print(f"FAIL error: {e}")
                time.sleep(2)

        if not found:
            print("FAIL no result found")
            failed += 1

    client.close()

    print(f"\n=== Summary ===")
    print(f"Success: {success}  Failed: {failed}")
    print(f"\nStored ({stored}):")
    for c, n in sorted(country_counts.items(), key=lambda x: -x[1]):
        print(f"  {c}: {n}")


if __name__ == "__main__":
    main()
